File: backend-javascript/03-backend-javascript-postgresql/src/modules/profession/profession.repository.js
import MockRepository from './profession.repository.mock.js';
import PgRepository from './profession.repository.pg.js';

class Repository {
  constructor(useDatabase) {
    this.repository = useDatabase
      ? new PgRepository()
      : new MockRepository();
  }

  async getItems() {
    return this.repository.getItems();
  }

  async getItemsCount() {
    return this.repository.getItemsCount();
  }

  async getItemById(id) {
    return this.repository.getItemById(id);
  }

  async createItem(data) {
    return this.repository.createItem(data);
  }

  async updateItem(id, data) {
    return this.repository.updateItem(id, data);
  }

  async deleteItem(id) {
    return this.repository.deleteItem(id);
  }

  async existsByName(name) {
    return await this.repository.existsByName(name);
  }
}

export default Repository;

File: mcp/java/backend-java/src/main/java/com/ganatan/modules/profession/ProfessionRepository.java
package com.ganatan.modules.profession;

import java.util.Map;
import java.util.Optional;

import com.ganatan.mocks.ProfessionRepositoryMock;
import com.ganatan.modules.profession.ProfessionRepositoryPg;

public class ProfessionRepository {

	private final Object repository;

	public ProfessionRepository(String useDbClient) {
		switch (useDbClient.toLowerCase()) {
		case "mock":
			this.repository = new ProfessionRepositoryMock();
			break;
		case "pg":
			this.repository = new ProfessionRepositoryPg();
			break;
		case "mysql":
			this.repository = new ProfessionRepositoryMock();
			break;
		default:
			this.repository = new ProfessionRepositoryMock();
			break;
		}
	}

	@SuppressWarnings("unchecked")
	public Map<String, Object> getItems(int offset, int limit) {
		System.out.println("00000000001:ProfessionRepository");
		if (repository instanceof ProfessionRepositoryPg pg) {
			System.out.println("00000000002:ProfessionRepository");
			return pg.getItems(offset, limit);
		}
		return ((ProfessionRepositoryMock) repository).getItems(offset, limit);
	}

	public Optional<Profession> getItemById(int id) {
		if (repository instanceof ProfessionRepositoryPg pg) {
			return pg.getItemById(id);
		}
		return ((ProfessionRepositoryMock) repository).getItemById(id);
	}

	public Profession createItem(Profession profession) {
		if (repository instanceof ProfessionRepositoryPg pg) {
			return pg.createItem(profession);
		}
		return ((ProfessionRepositoryMock) repository).createItem(profession);
	}

	public Optional<Profession> updateItem(int id, Profession updatedData) {
		if (repository instanceof ProfessionRepositoryPg pg) {
			return pg.updateItem(id, updatedData);
		}
		return ((ProfessionRepositoryMock) repository).updateItem(id, updatedData);
	}

	public boolean deleteItem(int id) {
		if (repository instanceof ProfessionRepositoryPg pg) {
			return pg.deleteItem(id);
		}
		return ((ProfessionRepositoryMock) repository).deleteItem(id);
	}

	public boolean existsByName(String name) {
		if (repository instanceof ProfessionRepositoryPg pg) {
			return pg.existsByName(name);
		}
		return ((ProfessionRepositoryMock) repository).existsByName(name);
	}
}
